### PR TITLE
Use returned radius attributes in order to tests them in rules

### PR DESCRIPTION
**Pull request by fdurand**
# Description
When you do radius authentication (radius source) you can use the radius attributes returned in the answer in order to use it to compute some rules.

# Impacts
Authentication flow

# Delete branch after merge
YES

# NEWS file entries

## Enhancements
* Use the attributes returned by a radius use source as attributes to compute the rules.

### CODE REVIEW

**Comment by dwlfrth** [comment]
@jrouzierinverse see if it conflicts with authentication sources rework. Let @fdurand know and adjust.
Depending on the work, we aim at 7.2

**Comment by jrouzierinverse** [comment]
There are no conflicts

File: lib/pf/Authentication/Source/RADIUSSource.pm [hunk 1]
```diff
@@ -167,15 +180,23 @@ sub check_radius_password {
...
     $params->{'username'} = $params->{'stripped_user_name'} if (defined($params->{'stripped_user_name'} ) && $params->{'stripped_user_name'} ne '' && isenabled($self->{'stripped_user_name'}));
     my $username =  $params->{'username'};
+
     foreach my $condition (@{ $own_conditions }) {
         if ($condition->{'attribute'} eq "username") {
             if ( $condition->matches("username", $username) ) {
                 push(@{ $matching_conditions }, $condition);
             }
         }
+        for my $attribute (@{ $extra->{attributes}} ) {
```
Comment (jrouzierinverse) [inline lib/pf/Authentication/Source/RADIUSSource.pm]: Add a check to make sure extra is defined

**Review by jrouzierinverse — CHANGES_REQUESTED** [review]
Extra is not used for matching here
html/captive-portal/lib/captiveportal/PacketFence/Controller/DeviceRegistration.pm

Extra is not save in the session here
html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm

In create_local_account extra is not used for matching there

html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication.pm

Should we use extra here even though these sources do not return the extra field
html/captive-portal/lib/captiveportal/PacketFence/Controller/Activate/Email.pm
html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Sponsor.pm

**Comment by fdurand** [comment]
Extra is save in the session here
html/captive-portal/lib/captiveportal/PacketFence/DynamicRouting/Module/Authentication/Login.pm +207

Also not sure we have to use the extra field in email and sponsor. Right now only radius source support this extra field.

File: lib/pf/Authentication/Source/RADIUSSource.pm [hunk 2]
```diff
@@ -167,15 +180,25 @@ sub check_radius_password {
...
     $params->{'username'} = $params->{'stripped_user_name'} if (defined($params->{'stripped_user_name'} ) && $params->{'stripped_user_name'} ne '' && isenabled($self->{'stripped_user_name'}));
     my $username =  $params->{'username'};
+
     foreach my $condition (@{ $own_conditions }) {
         if ($condition->{'attribute'} eq "username") {
             if ( $condition->matches("username", $username) ) {
                 push(@{ $matching_conditions }, $condition);
             }
         }
+        if defined($extra->{attributes}) {
```
Comment (jrouzierinverse) [inline lib/pf/Authentication/Source/RADIUSSource.pm]: That would not compile you forgot to put () around your condition. 
Yes switching between go and perl is annoying
Just check if $extra is defined
if (defined $extra) {

**Review by jrouzierinverse — COMMENTED** [review]
Has compile errors

**Review by jrouzierinverse — APPROVED** [review]

**Pull request merged**